File: apps/slate-mac/Sources/SlateMac/Graph/GraphConfigStore.swift
// Copyright (C) 2026 Cory Joseph
// SPDX-License-Identifier: AGPL-3.0-or-later

import Foundation

/// `.slate/graph.json` reader + atomic writer (Milestone P, P2-4 #560).
///
/// Write discipline mirrors the O-5 `history_prefs.rs` pattern: atomic
/// temp-file + rename (a kill mid-write never leaves torn JSON),
/// unknown future top-level keys preserved verbatim on rewrite
/// (forward-compat), and **refuse-to-clobber-unparseable** — if the file
/// exists but isn't a JSON object, the write throws rather than
/// destroying whatever a newer Slate (or a human) put there.
///
/// SINGLE-WRITER: only the Mac app writes `graph.json`, so — unlike
/// `prefs.json` (which the Rust history-prefs writer co-writes and
/// therefore flocks) — no lock is taken. A separate file from
/// `prefs.json` precisely to avoid contending on its `.lock`.
struct GraphConfigStore {
    let vaultRoot: URL

    var configURL: URL {
        vaultRoot.appendingPathComponent(".slate", isDirectory: true)
            .appendingPathComponent("graph.json", isDirectory: false)
    }

    /// Read the config. Missing file → `.default`. Malformed JSON throws
    /// (the caller surfaces it and leaves the file untouched).
    func read() throws -> GraphConfig {
        let url = configURL
        guard FileManager.default.fileExists(atPath: url.path) else { return .default }
        let data: Data
        do {
            data = try Data(contentsOf: url)
        } catch {
            throw PrefsJsonStoreError.readFailed(path: url.path, reason: error.localizedDescription)
        }
        // The codec is core's (W6-2 PR 0b, contracts doc 0b-12): a non-object
        // root, a version that cannot be classified (0b-D8) and a FORWARD
        // version all surface as parse failures, so the caller marks the
        // config read-only and never rewrites the file.
        let text = String(decoding: data, as: UTF8.self)
        do {
            return try graphConfigDecode(json: text).config
        } catch let error as GraphConfigError {
            throw PrefsJsonStoreError.parseFailed(path: url.path, reason: Self.reason(of: error))
        }
    }

    /// The host-facing reason for each core error (the adapter's mapping).
    static func reason(of error: GraphConfigError) -> String {
        switch error {
        case .Unparseable(let reason): return reason
        case .NewerVersion(let version):
            return "graph.json is a newer version (\(version)); not downgrading"
        }
    }

    /// Write the config, preserving unknown top-level keys and refusing
    /// to clobber an unparseable existing file. Atomic via temp+rename.
    func write(_ config: GraphConfig) throws {
        try ensureSlateDirExists()
        let url = configURL

        var existing: String?
        if FileManager.default.fileExists(atPath: url.path) {
            // Read the existing file THROWINGLY: a file that exists but can't
            // be read (permissions, transient I/O) must NOT be treated like a
            // missing file and overwritten — that would clobber whatever it
            // holds. Refuse instead (finding 2). `try?` here was the bug.
            do {
                existing = String(decoding: try Data(contentsOf: url), as: UTF8.self)
            } catch {
                throw PrefsJsonStoreError.writeFailed(
                    path: url.path,
                    reason: "existing graph.json is unreadable; refusing to overwrite: "
                        + error.localizedDescription)
            }
        }

        // The merge policy is core's (W6-2 PR 0b, contracts doc 0b-12): an
        // unparseable existing file is never clobbered, an unclassifiable or
        // newer version never downgraded, every unknown top-level key
        // preserved, the bytes canonical (0b-D5).
        let text: String
        do {
            text = try graphConfigEncode(config: config, existingJson: existing)
        } catch let error as GraphConfigError {
            throw PrefsJsonStoreError.writeFailed(path: url.path, reason: Self.reason(of: error))
        }
        let out = Data(text.utf8)
        do {
            try out.write(to: url, options: .atomic)
        } catch {
            throw PrefsJsonStoreError.writeFailed(path: url.path, reason: error.localizedDescription)
        }
    }

    private func ensureSlateDirExists() throws {
        let dir = vaultRoot.appendingPathComponent(".slate", isDirectory: true)
        guard !FileManager.default.fileExists(atPath: dir.path) else { return }
        do {
            try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        } catch {
            throw PrefsJsonStoreError.writeFailed(path: dir.path, reason: error.localizedDescription)
        }
    }

}

/// Serializes every `graph.json` write app-wide (Milestone P, P2-4 #560,
/// finding 3). The atomic temp+rename already prevents TORN JSON, but two
/// debounced saves whose read-merge-write cycles overlap could still lose
/// an update (both read the same base, the slower rename wins with the
/// older payload). Funnelling all writes through one actor makes the
/// read-merge-write atomic w.r.t. other writes in THIS process — a `write`
/// runs to completion before the next queued one starts.
///
/// MONOTONIC: each write carries a per-vault, strictly-increasing
/// `generation`; the actor records the newest generation it has written
/// per vault and DROPS any write whose generation is older. So even if the
/// actor executor delivers two queued writes out of order, a stale snapshot
/// can never clobber a newer one (finding 3 round-3 — the check must live
/// in the actor, not after the write on the caller side).
///
/// SINGLE-INSTANCE scope: Slate's Mac app is single-instance, so this is
/// the only writer of any vault's `graph.json`; cross-PROCESS contention
/// (two app instances on one vault) is out of scope and not locked (unlike
/// `prefs.json`, which the Rust core co-writes and therefore flocks).
/// Persistence is best-effort — a failed write (e.g. the refuse-to-clobber
/// guard) is swallowed here; the caller's `graphConfigWritable` gate is the
/// authoritative protection.
actor GraphConfigWriter {
    static let shared = GraphConfigWriter()

    /// Newest generation actually written per vault (monotonic gate).
    private var written: [URL: Int] = [:]

    func write(vault: URL, config: GraphConfig, generation: Int) {
        // Reject a superseded snapshot regardless of delivery order.
        if let seen = written[vault], generation < seen { return }
        written[vault] = generation
        // Best-effort persistence, but LOG failures rather than swallowing
        // them (Codoki review): a permission / disk-full / refuse-to-clobber
        // error must not crash or propagate (the caller's `graphConfigWritable`
        // gate is the authoritative protection), yet a silent drop makes a
        // field "my graph settings won't save" report undiagnosable.
        do {
            try GraphConfigStore(vaultRoot: vault).write(config)
        } catch {
            // Log the FULL vault path so the failing vault is uniquely
            // identifiable (two vaults can share a folder name) — matching
            // the AppState persistence-log convention. Fixed format string:
            // a path / error can legally contain `%`, which NSLog would
            // otherwise read as an unsupplied format specifier (corrupting
            // the message, e.g. `%@` → `(null)`).
            let message = "Failed to persist graph.json for vault '\(vault.path)': \(error)"
            NSLog("%@", message)
        }
    }
}

File: apps/slate-windows/src/SlateWindows/Graph/GraphConfigStore.cs
// Copyright (C) 2026 Cory Joseph
// SPDX-License-Identifier: AGPL-3.0-or-later

using System.Diagnostics;
using System.IO;
using System.Text;
using uniffi.slate_uniffi;

namespace SlateWindows.Graph;

/// <summary>What the store's read answered (rule W, Term W7's decode
/// arms): the config in memory, whether the file may be written back,
/// and — when the default stood in — the reason.</summary>
internal sealed record GraphConfigLoad(GraphConfig Config, bool Writable, string? Failure);

/// <summary>
/// W6-2 PR C (#746), contract C-10: the host I/O 0bD-3 leaves to the host
/// — the mac's <c>GraphConfigStore.swift</c> twin — and nothing else.
/// READ: a missing file is the default and writable; an unreadable,
/// unparseable or newer-version file is the default, NOT writable, the
/// file untouched, the reason logged (Term W7). WRITE: the existing text
/// read THROWING (an unreadable file refuses the write rather than
/// clobbering it), core's merge through <c>graph_config_encode</c> (an
/// unparseable existing file never clobbered, a newer version never
/// downgraded, unknown keys preserved, the bytes canonical — 0b-12), the
/// bytes to a temporary file beside the target and one
/// <c>File.Move</c> over it, <c>.slate</c> created when missing. The
/// writer census (C-15 xv) walls this class as the ONE writer of the
/// file and this constant as the one place its name appears.
/// </summary>
internal sealed class GraphConfigStore
{
    /// <summary>The file's name, in the one place it appears (C-15 xv).</summary>
    internal const string FileName = "graph.json";

    private readonly string _path;

    public GraphConfigStore(string vaultRoot)
    {
        ArgumentNullException.ThrowIfNull(vaultRoot);
        _path = Path.Combine(vaultRoot, ".slate", FileName);
    }

    public string FilePath => _path;

    /// <summary>Test seam: the existing text's read, in place of the file's;
    /// a fact makes it throw to prove the write refuses rather than clobbers.</summary>
    internal Func<string, string>? ReadExistingForTests { get; set; }

    /// <summary>Term W7's decode arms; the file is never rewritten on a read.</summary>
    public GraphConfigLoad Read()
    {
        if (!File.Exists(_path))
        {
            return new GraphConfigLoad(SlateUniffiMethods.GraphConfigDefault(), Writable: true, Failure: null);
        }
        string text;
        try
        {
            text = File.ReadAllText(_path);
        }
        catch (Exception exception) when (exception is IOException or UnauthorizedAccessException)
        {
            string reason = $"{FileName} is unreadable: {exception.Message}";
            Trace.TraceWarning("graph config at '{0}' read as the default, read-only: {1}", _path, reason);
            return new GraphConfigLoad(SlateUniffiMethods.GraphConfigDefault(), Writable: false, Failure: reason);
        }
        try
        {
            return new GraphConfigLoad(SlateUniffiMethods.GraphConfigDecode(text).Config, Writable: true, Failure: null);
        }
        catch (GraphConfigException exception)
        {
            string reason = Reason(exception);
            Trace.TraceWarning("graph config at '{0}' read as the default, read-only: {1}", _path, reason);
            return new GraphConfigLoad(SlateUniffiMethods.GraphConfigDefault(), Writable: false, Failure: reason);
        }
    }

    /// <summary>The write: read-merge-write through core's codec, atomic
    /// over the target. Throws — an unreadable existing file, core's
    /// refusal, a filesystem failure — and the writer logs (rule W, Term
    /// W2); nothing here swallows.</summary>
    public void Write(GraphConfig config)
    {
        ArgumentNullException.ThrowIfNull(config);
        string? directory = Path.GetDirectoryName(_path);
        if (!string.IsNullOrEmpty(directory))
        {
            Directory.CreateDirectory(directory);
        }
        string? existing = null;
        if (File.Exists(_path))
        {
            // THROWING (the mac's finding 2): a file that exists but cannot
            // be read must not be treated like a missing one and overwritten.
            existing = ReadExistingForTests is { } read ? read(_path) : File.ReadAllText(_path);
        }
        string text = SlateUniffiMethods.GraphConfigEncode(config, existing);
        string temporary = $"{_path}.{Guid.NewGuid():N}.tmp";
        try
        {
            File.WriteAllText(temporary, text, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false));
            File.Move(temporary, _path, overwrite: true);
        }
        finally
        {
            SafeFile.TryDelete(temporary);
        }
    }

    /// <summary>The host-facing reason for each core refusal (the mac's
    /// adapter mapping): the parser's message, or the version not downgraded.</summary>
    internal static string Reason(GraphConfigException exception)
    {
        ArgumentNullException.ThrowIfNull(exception);
        return exception is GraphConfigException.NewerVersion
            ? $"{FileName} is a newer version; not downgrading ({exception.Message})"
            : exception.Message;
    }
}
